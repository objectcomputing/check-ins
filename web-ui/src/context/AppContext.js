import React, { useEffect, useReducer, useMemo } from "react";
import { createSelector } from 'reselect';
import { getCurrentUser, updateMember, getAllMembers } from "../api/member";
import { getCheckinByMemberId, createCheckin } from "../api/checkins";
import { BASE_API_URL } from "../api/api";
import axios from "axios";

export const MY_PROFILE_UPDATE = "@@check-ins/update_profile";
export const UPDATE_USER_BIO = "@@check-ins/update_bio";
export const UPDATE_CHECKINS = "@@check-ins/update_checkins";
export const UPDATE_INDEX = "@@check-ins/update_index";
export const UPDATE_TOAST = "@@check-ins/update_toast";
export const UPDATE_CURRENT_CHECKIN = "@@check-ins/update_current_checkin";
export const UPDATE_TEAMS = "@@check-ins/update_teams";
export const UPDATE_MEMBER_PROFILES = "@@check-ins/update_member_profiles";
export const UPDATE_TEAM_MEMBERS = "@@check-ins/update_team_members";
export const UPDATE_SELECTED_PROFILE = "@@check-ins/update_selected_profile";
const SET_CSRF = "@@check-ins/update_csrf";

const AppContext = React.createContext();

const reducer = (state, action) => {
  switch (action.type) {
    case MY_PROFILE_UPDATE:
      state.userProfile = action.payload;
      break;
    case UPDATE_USER_BIO:
      state.userProfile.memberProfile.bioText = action.payload;
      updateMember(state.userProfile.memberProfile);
      break;
    case UPDATE_CHECKINS:
      state.checkins = action.payload;
      //sort by date
      state.checkins.sort(function (a, b) {
        return new Date(...a.checkInDate) - new Date(...b.checkInDate);
      });
      state.currentCheckin = state.checkins[state.checkins.length - 1];
      break;
    case SET_CSRF:
      state.csrf = action.payload;
      break;
    case UPDATE_TOAST:
      state.toast = action.payload;
      break;
    case UPDATE_TEAMS:
      state.teams = action.payload;
      break;
    case UPDATE_MEMBER_PROFILES:
      state.memberProfiles = action.payload;
      break;
    case UPDATE_TEAM_MEMBERS:
      state.teamMembers = action.payload;
      break;
    case UPDATE_CURRENT_CHECKIN:
      state.currentCheckin = action.payload;
      break;
    case UPDATE_SELECTED_PROFILE:
      const { payload } = action;
      if (state.selectedProfile !== payload) {
        state.selectedProfile = payload;
        state.currentCheckin = payload ? payload.checkIn : {};
      }
      if (payload === undefined) {
        state.checkins = [];
      }
      break;
    default:
  }
  return { ...state };
};

const initialState = {
  checkins: [],
  csrf: undefined,
  currentCheckin: {},
  teams: [],
  memberProfiles: [],
  index: 0,
  selectedProfile: undefined,
  toast: {
    severity: "",
    toast: "",
  },
  userProfile: undefined,
};

const getCheckins = async (id, pdlId, date, dispatch, csrf) => {
  const res = await getCheckinByMemberId(id, csrf);
  let data =
    res.payload && res.payload.data && res.payload.status === 200 && !res.error
      ? res.payload.data
      : null;
  if (data && data.length > 0) {
    const allComplete = data.every((checkin) => checkin.completed === true);
    if (allComplete) {
      const prevCheckinDate = data[data.length - 1].checkInDate;
      if (pdlId) {
        const res = await createCheckin({
          teamMemberId: id,
          pdlId: pdlId,
          checkInDate: date(3, prevCheckinDate),
          completed: false,
        });
        const checkin =
          res.payload && res.payload.data && !res.error
            ? res.payload.data
            : null;
        data.push(checkin);
      }
    }
  } else if (data && data.length === 0) {
    if (pdlId) {
      const res = await createCheckin({
        teamMemberId: id,
        pdlId: pdlId,
        checkInDate: date(1),
        completed: false,
      });
      const checkin =
        res.payload && res.payload.data && !res.error ? res.payload.data : null;
      data = [checkin];
    }
  }
  //without this check you get infinite checkin calls
  if (data.length > 0) {
    dispatch({ type: UPDATE_CHECKINS, payload: data });
  }
};

const AppContextProvider = (props) => {
  const [state, dispatch] = useReducer(reducer, initialState);
  const memberProfile =
    state && state.userProfile && state.userProfile.memberProfile
      ? state.userProfile.memberProfile
      : undefined;
  const id = memberProfile ? memberProfile.id : undefined;
  const selectedProfile = state && state.selectedProfile;
  const selectedId = selectedProfile ? selectedProfile.id : undefined;


  const { csrf } = state;

  useEffect(() => {
    const getCsrf = async () => {
      if (!csrf) {
        const res = await axios({
          url: `${BASE_API_URL}/csrf/cookie`,
          responseType: "text",
          withCredentials: true,
        });
        if (res && res.data) {
          dispatch({ type: SET_CSRF, payload: res.data });
        }
      }
    };
    getCsrf();
  }, [csrf]);

  useEffect(() => {
    const updateUserProfile = async () => {
      let res = await getCurrentUser(csrf);
      let profile =
        res.payload && res.payload.data && !res.error
          ? res.payload.data
          : undefined;

      if (profile) {
        dispatch({ type: MY_PROFILE_UPDATE, payload: profile });
      }
    };
    if (csrf) {
      updateUserProfile();
    }
  }, [csrf]);

  const date = (months, prevCheckinDate) => {
    let currentMonth = new Date().getMonth() + 1;
    let newDate = prevCheckinDate ? new Date(...prevCheckinDate) : new Date();
    if (prevCheckinDate) {
      let prevCheckinMonth = prevCheckinDate.getMonth() + 1;
      newDate.setMonth(
        newDate.getMonth() + currentMonth - prevCheckinMonth >= 3 ? 1 : months
      );
    } else {
      newDate.setMonth(newDate.getMonth() + months);
    }
    const year = newDate.getFullYear();
    const month = newDate.getMonth() + 1;
    const day = newDate.getDate();
    const hours = newDate.getHours();
    const minutes = newDate.getMinutes();
    const dateTimeArray = [year, month, day, hours, minutes, 0];
    return dateTimeArray;
  };

  useEffect(() => {
    async function getMemberProfiles() {
      let res = await getAllMembers(csrf);
      let profiles =
        res.payload && res.payload.data && !res.error
          ? res.payload.data
          : undefined;

      if (profiles) {
        dispatch({ type: UPDATE_MEMBER_PROFILES, payload: profiles });
      }
    }
    if (csrf) {
      getMemberProfiles();
    }
  }, [csrf]);

  useEffect(() => {
    if (selectedId && csrf) {
      getCheckins(selectedId, id, date, dispatch, csrf);
    }
  }, [csrf, selectedId, id]);

  const value = useMemo(() => {
    return { state, dispatch };
  }, [state]);
  return (
    <AppContext.Provider value={props.value || value}>
      {props.children}
    </AppContext.Provider>
  );
};

export const selectMemberProfiles = state => state.memberProfiles;
export const selectTeamMembers = state => state.teamMembers;

export const selectProfileMap = createSelector(
  selectMemberProfiles,
  (memberProfiles) => {
    if (memberProfiles && memberProfiles.length) {
      memberProfiles = memberProfiles.reduce((mappedById, profile) => {
        mappedById[profile.id] = profile;
        return mappedById;
      }, {});
    }
    return memberProfiles;
  }
);

export const selectMembersByTeamId = id => createSelector(
  selectTeamMembers,
  (teamMembers) => {
    let members = [];
    if (teamMembers && teamMembers.length) {
      members = teamMembers.filter((member) => member.teamid === id);
    }
    return members;
  }
);

export const selectMemberProfilesByTeamId = id => createSelector(
  selectMembersByTeamId(id),
  selectProfileMap,
  (members, profileMap) => members.map((member) => {
      return { ...profileMap[member.memberid], ...member };
    })
);

export { AppContext, AppContextProvider };
