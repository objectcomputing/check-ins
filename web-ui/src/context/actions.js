export const MY_PROFILE_UPDATE = "@@check-ins/update_profile";
export const UPDATE_USER_BIO = "@@check-ins/update_bio";
export const UPDATE_CHECKINS = "@@check-ins/update_checkins";
export const UPDATE_CHECKIN = "@@check-ins/update_checkin";
export const ADD_CHECKIN = "@@check-ins/add_checkin";
export const UPDATE_TOAST = "@@check-ins/update_toast";
export const UPDATE_MEMBER_SKILLS = "@@check-ins/update_member_skills";
export const DELETE_MEMBER_SKILL = "@@check-ins/delete_member_skill";
export const ADD_MEMBER_SKILL = "@@check-ins/add_member_skill";
export const UPDATE_TEAMS = "@@check-ins/update_teams";
export const UPDATE_MEMBER_PROFILES = "@@check-ins/update_member_profiles";
export const UPDATE_TEAM_MEMBERS = "@@check-ins/update_team_members";
export const ADD_SKILL = "@@check-ins/add-skill";
export const DELETE_SKILL = "@@check-ins/delete-skill";
export const UPDATE_SKILL = "@@check-ins/update_skill";
export const UPDATE_SKILLS = "@@check-ins/update_skills";
export const ADD_TEAM = "@@check-ins/add_team";
export const SET_CSRF = "@@check-ins/update_csrf";
export const SET_ROLES = "@@check-ins/set_roles";
