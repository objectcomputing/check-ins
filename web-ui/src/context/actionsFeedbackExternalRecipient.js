export const SET_CSRF = '@@check-ins/update_csrf';
export const UPDATE_TOAST = '@@check-ins/update_toast';

