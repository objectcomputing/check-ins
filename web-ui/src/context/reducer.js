import {
  ADD_CHECKIN,
  ADD_MEMBER_SKILL,
  ADD_SKILL,
  ADD_TEAM,
  DELETE_MEMBER_PROFILE,
  DELETE_MEMBER_SKILL,
  DELETE_SKILL,
  MY_PROFILE_UPDATE,
  SET_CSRF,
  SET_ROLES,
  SET_SELECTED_MEMBER,
  UPDATE_CHECKIN,
  UPDATE_CHECKINS,
  UPDATE_MEMBER_PROFILES,
  UPDATE_MEMBER_SKILLS,
  UPDATE_SKILL,
  UPDATE_SKILLS,
  UPDATE_TEAMS,
  UPDATE_TEAM_MEMBERS,
  UPDATE_TOAST,
  UPDATE_USER_BIO,
} from "./actions";

export const initialState = {
  checkins: [],
  csrf: undefined,
  index: 0,
  memberProfiles: [],
  memberSkills: [],
  selectedMember: undefined,
  skills: [],
  teams: [],
  toast: {
    severity: "",
    toast: "",
  },
  userProfile: undefined,
};

export const reducer = (state, action) => {
  switch (action.type) {
    case MY_PROFILE_UPDATE:
      state.userProfile = action.payload;
      break;
    case UPDATE_USER_BIO:
      state.userProfile.memberProfile.bioText = action.payload;
      break;
    case ADD_CHECKIN:
      state.checkins = [...state.checkins, action.payload];
      break;
    case UPDATE_CHECKINS:
      if (state?.checkins?.length > 0) {
        state.checkins = [...state.checkins];
        action.payload.forEach((checkin) => {
          const checkInIndex = state.checkins.findIndex(
            (current) => current.id === checkin.id
          );
          if (checkInIndex >= 0) {
            state.checkins[checkInIndex] = checkin;
          } else {
            state.checkins.push(checkin);
          }
        });
      } else {
        state.checkins = action.payload;
      }
      break;
    case UPDATE_CHECKIN:
      state.checkins = [...state.checkins];
      const checkInIndex = state.checkins.findIndex(
        (checkin) => checkin.id === action.payload.id
      );
      if (checkInIndex >= 0) {
        state.checkins[checkInIndex] = action.payload;
      } else {
        state.checkins.push(action.payload);
      }
      break;
    case ADD_SKILL:
      state.skills = [...state.skills, action.payload];
      break;
    case DELETE_SKILL:
      state.skills = state.skills.filter(
        (skill) => skill.id !== action.payload
      );
      break;
    case UPDATE_SKILL:
      state.skills = [...state.skills];
      const index = state.skills.findIndex(
        (skill) => skill.id === action.payload.id
      );
      state.skills[index] = action.payload;
      break;
    case UPDATE_SKILLS:
      state.skills = action.payload;
      break;
    case SET_CSRF:
      state.csrf = action.payload;
      break;
    case UPDATE_TOAST:
      state.toast = action.payload;
      break;
    case ADD_TEAM:
      state.teams = [...state.teams, action.payload];
      //sort by name
      state.teams.sort((a, b) => a.name.localeCompare(b.name));
      break;
    case UPDATE_TEAMS:
      state.teams = action.payload;
      //sort by name
      state.teams.sort((a, b) => a.name.localeCompare(b.name));
      break;
    case UPDATE_MEMBER_PROFILES:
      state.memberProfiles = action.payload;
      break;
    case UPDATE_TEAM_MEMBERS:
      state.teamMembers
        ? (state.teamMembers = [...state.teamMembers, action.payload])
        : (state.teamMembers = action.payload);
      break;
    case UPDATE_MEMBER_SKILLS:
      state.memberSkills = action.payload;
      break;
    case DELETE_MEMBER_SKILL:
      state.memberSkills = [
        ...state.memberSkills.filter(
          (mSkill) => mSkill.skillid !== action.payload
        ),
      ];
      break;
    case DELETE_MEMBER_PROFILE:
      state.memberProfiles = [
        ...state.memberProfiles.filter(
          (profile) => profile.id !== action.payload
        ),
      ];
      break;
    case ADD_MEMBER_SKILL:
      state.memberSkills = [...state.memberSkills, action.payload];
      break;
    case SET_ROLES:
      state.roles = action.payload;
      break;
    case SET_SELECTED_MEMBER:
      state.selectedMember = action.payload;
      break;
    default:
  }
  return { ...state };
};
