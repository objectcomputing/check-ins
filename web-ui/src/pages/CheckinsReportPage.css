.MuiFormControl-root {
  margin-left: 1rem;
}
