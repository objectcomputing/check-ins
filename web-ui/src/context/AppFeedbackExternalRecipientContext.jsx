import React, { useEffect, useReducer, useMemo } from 'react';
import {reducerFeedbackExternalRecipient, initialState} from "./reducerFeedbackExternalRecilpients.js";
import {BASE_API_URL} from "../api/apiFeedbackExternalRecipient.js";
import {SET_CSRF} from "./actionsFeedbackExternalRecipient.js";

const AppFeedbackExternalRecipientContext = React.createContext();

function getSessionCookieValue(name) {
  const cookies = document?.cookie?.split(';');
  for (let i = 0; i < cookies.length; i++) {
    const cookie = cookies[i].trim();
    if (cookie.startsWith(name + '=')) {
      return decodeURIComponent(cookie.substring(name.length + 1));
    }
  }
  return null;
}

const AppFeedbackExternalRecipientContextProvider = props => {
  console.log("AppFeedbackExternalRecipientContextProvider props: ", props);
  const [state, dispatch] = useReducer(
      reducerFeedbackExternalRecipient,
    props?.value?.state || initialState
  );

  const {
    csrf,
  } = state;
  const url = `${BASE_API_URL}/csrf/cookie`;

  useEffect(() => {

    console.log("AppFeedbackExternalRecipientContextProvider useEffect, csrf");

    const getCsrf = async () => {
      if (!csrf) {
        const payload = getSessionCookieValue('_csrf');
        if (payload) {
          dispatch({ type: SET_CSRF, payload });
        } else {
          const res = await fetch(url, {
            responseType: 'text',
            credentials: 'include'
          });
          if (res && res.ok) {
            dispatch({ type: SET_CSRF, payload: await res.text() });
          }
        }
      }
    };
    getCsrf();
  }, [csrf]);

  const value = useMemo(() => {
    console.log("AppFeedbackExternalRecipientContextProvider useMemo, state: ", state);
    console.log("AppFeedbackExternalRecipientContextProvider useMemo, dispatch: ", dispatch);
    return { state, dispatch };
  }, [state]);

  return (
    <AppFeedbackExternalRecipientContext.Provider value={props.value || value}>
      {props.children}
    </AppFeedbackExternalRecipientContext.Provider>
  );
};

export { AppFeedbackExternalRecipientContext, AppFeedbackExternalRecipientContextProvider };
