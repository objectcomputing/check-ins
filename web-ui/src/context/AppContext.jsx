import React, { useEffect, useReducer, useMemo } from 'react';
import { reducer, initialState } from './reducer';
import { getCheckins, getAllCheckinsForAdmin } from './thunks';
import {
  MY_PROFILE_UPDATE,
  SET_CSRF,
  SET_ROLES,
  SET_USER_ROLES,
  UPDATE_GUILDS,
  UPDATE_MEMBER_SKILLS,
  UPDATE_MEMBER_PROFILES,
  UPDATE_TERMINATED_MEMBERS,
  UPDATE_SKILLS,
  UPDATE_CERTIFICATIONS,
  UPDATE_TEAMS,
  UPDATE_PEOPLE_LOADING,
  UPDATE_TEAMS_LOADING
} from './actions';
import {
  getCurrentUser,
  getAllMembers,
  getAllTerminatedMembers
} from '../api/member';
import {
  selectCanViewCheckinsPermission,
  selectCanViewTerminatedMembers,
} from './selectors';
import { getAllRoles, getAllUserRoles } from '../api/roles';
import { getMemberSkills } from '../api/memberskill';
import { BASE_API_URL } from '../api/api';
import { getAllGuilds } from '../api/guild';
import { getSkills } from '../api/skill';
import { getAllTeams } from '../api/team';
import {getCertifications} from "../api/certification.js";

const AppContext = React.createContext();

function getSessionCookieValue(name) {
  const cookies = document?.cookie?.split(';');
  for (let i = 0; i < cookies.length; i++) {
    const cookie = cookies[i].trim();
    if (cookie.startsWith(name + '=')) {
      return decodeURIComponent(cookie.substring(name.length + 1));
    }
  }
  return null;
}

const AppContextProvider = props => {
  const [state, dispatch] = useReducer(
    reducer,
    props?.value?.state || initialState
  );
  const userProfile =
    state && state.userProfile ? state.userProfile : undefined;
  const memberProfile =
    userProfile && userProfile.memberProfile
      ? userProfile.memberProfile
      : undefined;

  const id = memberProfile ? memberProfile.id : undefined;
  const pdlId = memberProfile ? memberProfile.pdlId : undefined;
  const {
    csrf,
    guilds,
    teams,
    memberSkills,
    memberProfiles,
    checkins,
    skills,
    certifications,
    roles,
    userRoles
  } = state;
  const url = `${BASE_API_URL}/csrf/cookie`;
  useEffect(() => {
    const getCsrf = async () => {
      if (!csrf) {
        const payload = getSessionCookieValue('_csrf');
        if (payload) {
          dispatch({ type: SET_CSRF, payload });
        } else {
          const res = await fetch(url, {
            responseType: 'text',
            credentials: 'include'
          });
          if (res && res.ok) {
            dispatch({ type: SET_CSRF, payload: await res.text() });
          }
        }
      }
    };
    getCsrf();
  }, [csrf]);

  useEffect(() => {
    async function getGuilds() {
      let res = await getAllGuilds(csrf);
      let data =
        res.payload &&
        res.payload.data &&
        res.payload.status === 200 &&
        !res.error
          ? res.payload.data
          : null;
      if (data) {
        dispatch({ type: UPDATE_GUILDS, payload: data });
      }
    }
    if (csrf && !guilds) {
      getGuilds();
    }
  }, [csrf, guilds]);

  useEffect(() => {
    async function getTeams() {
      let res = await getAllTeams(csrf);
      let data =
        res.payload &&
        res.payload.data &&
        res.payload.status === 200 &&
        !res.error
          ? res.payload.data
          : null;
      if (data) {
        dispatch({ type: UPDATE_TEAMS, payload: data });
        dispatch({ type: UPDATE_TEAMS_LOADING });
      }
    }
    if (csrf && !teams) {
      dispatch({ type: UPDATE_TEAMS_LOADING });
      getTeams();
    }
  }, [csrf, teams, dispatch]);

  useEffect(() => {
    const updateUserProfile = async () => {
      let res = await getCurrentUser(csrf);
      let profile =
        res.payload && res.payload.data && !res.error
          ? res.payload.data
          : undefined;

      if (profile) {
        dispatch({ type: MY_PROFILE_UPDATE, payload: profile });
      }
    };
    if (csrf && !userProfile) {
      updateUserProfile();
    }
  }, [csrf, userProfile]);

  useEffect(() => {
    const getAllMemberSkills = async () => {
      const res = await getMemberSkills(csrf);
      const memberSkills =
        res && res.payload && res.payload.data ? res.payload.data : null;
      if (memberSkills) {
        dispatch({ type: UPDATE_MEMBER_SKILLS, payload: memberSkills });
      }
    };
    if (csrf && !memberSkills) {
      getAllMemberSkills();
    }
  }, [csrf, memberSkills]);

  useEffect(() => {
    async function getMemberProfiles() {
      let res = await getAllMembers(csrf);
      let profiles =
        res.payload && res.payload.data && !res.error
          ? res.payload.data
          : undefined;

      if (profiles) {
        dispatch({ type: UPDATE_MEMBER_PROFILES, payload: profiles });
        dispatch({ type: UPDATE_PEOPLE_LOADING, payload: false });
      }
    }
    async function getTerminatedMembers() {
      let res = await getAllTerminatedMembers(csrf);
      let profiles =
        res.payload && res.payload.data && !res.error
          ? res.payload.data
          : undefined;

      if (profiles) {
        dispatch({ type: UPDATE_TERMINATED_MEMBERS, payload: profiles });
      }
    }
    if (csrf && userProfile && !memberProfiles) {
      dispatch({ type: UPDATE_PEOPLE_LOADING, payload: true });
      getMemberProfiles();
      if (selectCanViewTerminatedMembers(state)) {
        getTerminatedMembers();
      }
    }
  }, [csrf, userProfile, memberProfiles]);

  useEffect(() => {
    function getAllTheCheckins() {
      if (
        userProfile &&
        userProfile.permissions?.some(p =>
          p?.permission?.includes('CAN_VIEW_CHECKINS_REPORT')
        ) &&
        id &&
        csrf
      ) {
        getAllCheckinsForAdmin(dispatch, csrf);
      } else if (id && csrf && selectCanViewCheckinsPermission(state)) {
        getCheckins(id, pdlId, dispatch, csrf);
      }
    }
    if (csrf && !checkins) {
      getAllTheCheckins();
    }
  }, [csrf, pdlId, id, userProfile, checkins]);

  useEffect(() => {
    const getAllSkills = async () => {
      const res = await getSkills(csrf);
      const data =
        res &&
        res.payload &&
        res.payload.data &&
        res.payload.status === 200 &&
        !res.error
          ? res.payload.data
          : null;
      if (data && data.length > 0) {
        dispatch({ type: UPDATE_SKILLS, payload: data });
      }
    };
    if (csrf && !skills) {
      getAllSkills();
    }
  }, [csrf, skills]);

  useEffect(() => {
    const getAllCertifications = async () => {
      const res = await getCertifications(csrf);
      const data =
          res &&
          res.payload &&
          res.payload.data &&
          res.payload.status === 200 &&
          !res.error
              ? res.payload.data
              : null;
      if (data && data.length > 0) {
        dispatch({ type: UPDATE_CERTIFICATIONS, payload: data });
      }
    };
    if (csrf && !certifications) {
      getAllCertifications();
    }
  }, [csrf, certifications]);

  useEffect(() => {
    const getRoles = async () => {
      const res = await getAllRoles(csrf);
      const data =
        res &&
        res.payload &&
        res.payload.data &&
        res.payload.status === 200 &&
        !res.error
          ? res.payload.data
          : null;
      if (data && Array.isArray(data) && data.length > 0) {
        dispatch({ type: SET_ROLES, payload: data });
      }
    };
    if (csrf && !roles) {
      getRoles();
    }
  }, [csrf, roles]);

  useEffect(() => {
    const getUserRoles = async () => {
      // make call to the API
      let res = await getAllUserRoles(csrf);
      return res.payload &&
        res.payload.data &&
        res.payload.status === 200 &&
        !res.error
        ? res.payload.data
        : null;
    };

    if (csrf && !userRoles) {
      getUserRoles().then(userRoles => {
        dispatch({ type: SET_USER_ROLES, payload: userRoles });
      });
    }
  }, [csrf, userRoles]);

  const value = useMemo(() => {
    return { state, dispatch };
  }, [state]);
  return (
    <AppContext.Provider value={props.value || value}>
      {props.children}
    </AppContext.Provider>
  );
};

export { AppContext, AppContextProvider };
