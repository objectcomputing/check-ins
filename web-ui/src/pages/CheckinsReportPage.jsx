import React, { useContext, useEffect, useState } from "react";

import { AppContext } from "../context/AppContext";
import CheckinReport from "../components/Reports/CheckinReport";
import {
  selectCheckinPDLS,
  selectTeamMembersWithCheckinPDL,
} from "../context/selectors";

import { TextField } from "@material-ui/core";
import Autocomplete from "@material-ui/lab/Autocomplete";

import "./CheckinsReportPage.css";

const CheckinsReportPage = () => {
  const { state } = useContext(AppContext);
  const pdls = selectCheckinPDLS(state).sort((a, b) => {
    const aPieces = a.name.split(" ").slice(-1);
    const bPieces = b.name.split(" ").slice(-1);
    return aPieces.toString().localeCompare(bPieces);
  });
  const [filteredPdls, setFilteredPdls] = useState(pdls);
  const [selectedPdls, setSelectedPdls] = useState([]);

  useEffect(() => {
    if (!pdls) return;
    pdls.map(
      (pdl) => (pdl.members = selectTeamMembersWithCheckinPDL(state, pdl.id))
    );
    setFilteredPdls(pdls);
  }, [pdls, filteredPdls, state]);

  const onPdlChange = (event, newValue) => {
    let extantPdls = filteredPdls || [];
    newValue.forEach((val) => {
      extantPdls = extantPdls.filter((pdl) => pdl.id !== val.id);
    });
    extantPdls = [...new Set(extantPdls)];
    newValue = [...new Set(newValue)];
    setFilteredPdls([...newValue]);
    setSelectedPdls(newValue);
  };

  return (
    <div>
      <div>
        <Autocomplete
          id="pdlSelect"
          multiple
          options={pdls}
          value={selectedPdls || []}
          onChange={onPdlChange}
          getOptionLabel={(option) => option.name}
          renderInput={(params) => (
            <TextField
              {...params}
              label="Select PDLs"
              placeholder="Choose which PDLs to display"
            />
          )}
        />
      </div>
      {selectedPdls.length
        ? selectedPdls.map((pdl) => <CheckinReport key={pdl.id} pdl={pdl} />)
        : filteredPdls.map((pdl) => <CheckinReport key={pdl.id} pdl={pdl} />)}
    </div>
  );
};

export default CheckinsReportPage;
