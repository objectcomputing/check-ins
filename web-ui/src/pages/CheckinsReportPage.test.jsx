import React from 'react';
import CheckinsReportPage from './CheckinsReportPage';
import { AppContextProvider } from '../context/AppContext';
import { BrowserRouter } from 'react-router-dom';

const userStateWithPermission = {
  state: {
    userProfile: {
      name: 'Mitch Hedberg',
      role: ['MEMBER'],
      permissions: [{ permission: 'CAN_VIEW_CHECKINS_REPORT' }],
    },
    terminatedMembers: [],
  }
};

it('renders correctly', () => {
  snapshot(
    <AppContextProvider value={userStateWithPermission}>
      <BrowserRouter>
        <CheckinsReportPage />
      </BrowserRouter>
    </AppContextProvider>
  );
});

it('renders an error if user does not have appropriate permission', () => {
  snapshot(
    <AppContextProvider>
      <BrowserRouter>
        <CheckinsReportPage />
      </BrowserRouter>
    </AppContextProvider>
  );
});
