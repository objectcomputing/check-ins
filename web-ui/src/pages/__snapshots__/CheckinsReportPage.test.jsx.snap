// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders an error if user does not have appropriate permission 1`] = `
<div>
  <h3>
    You do not have permission to view this page.
  </h3>
</div>
`;

exports[`renders correctly 1`] = `
<div>
  <div
    class="checkins-report-page"
  >
    <div
      class="MuiBox-root css-1l4w6pd"
    >
      <div
        aria-label="Basic button group"
        class="MuiButtonGroup-root MuiButtonGroup-text css-iv8nej-MuiButtonGroup-root"
        role="group"
      >
        <button
          aria-label="Previous quarter\`"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-mf1cb5-MuiButtonBase-root-MuiIconButton-root"
          data-mui-internal-clone-element="true"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ArrowBackIcon"
            focusable="false"
            style="font-size: 1.2em;"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <h6
          class="MuiTypography-root MuiTypography-h6 css-16dfr7g-MuiTypography-root"
        >
          <nobr>
            Q4 2024
          </nobr>
        </h6>
        <button
          aria-label="Next quarter\`"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-mf1cb5-MuiButtonBase-root-MuiIconButton-root"
          data-mui-internal-clone-element="true"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ArrowForwardIcon"
            focusable="false"
            style="font-size: 1.2em;"
            viewBox="0 0 24 24"
          >
            <path
              d="m12 4-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <dl
        class="MuiGrid-root MuiGrid-container checkins-report-page-dates css-11lq3yg-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        >
          <dt
            class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
          >
            Start of Quarter
          </dt>
          <dd
            class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
          >
            Tue Oct 01 2024
          </dd>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        >
          <dt
            class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
          >
            End of Quarter
          </dt>
          <dd
            class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
          >
            Tue Dec 31 2024
          </dd>
        </div>
      </dl>
    </div>
    <div
      class="checkins-report-page-reports"
    >
      <div
        class="team-member-map MuiBox-root css-0"
      >
        <div
          class="team-member-map-row MuiBox-root css-0"
        >
          <div
            class="team-member-map-row-summary MuiBox-root css-0"
          >
            <div
              class="MuiBox-root css-0"
              style="cursor: pointer;"
            >
              <h5
                class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
              >
                Member
              </h5>
            </div>
            <div
              class="team-member-map-summmary-content"
            >
              <hgroup />
              <div
                class="MuiBox-root css-0"
                style="cursor: pointer;"
              >
                <h5
                  class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
                >
                  PDL
                </h5>
              </div>
              <div
                class="MuiBox-root css-1mvoz6j"
              >
                <h5
                  class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
                >
                  Check-In Date
                </h5>
              </div>
              <div
                class="MuiBox-root css-0"
              >
                <h5
                  class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
                >
                  Status
                </h5>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
